stdlib/public/core/NormalizedCodeUnitIterator.swift

```diff
@@ -226,15 +226,96 @@ extension _NormalizedUTF8CodeUnitIterator: Sequence { }
 internal
 struct _NormalizedUTF16CodeUnitIterator: IteratorProtocol {
   internal typealias CodeUnit = UInt16
-  var segmentBuffer = _FixedArray16<CodeUnit>(allZeros:())
-  var overflowBuffer: [CodeUnit]? = nil
-  var normalizationBuffer: [CodeUnit]? = nil
+  var segmentBuffer = NormalizationBuffer()
+  var intermediateBuffer = NormalizationBuffer()
   var source: _SegmentSource
-
-  var segmentBufferIndex = 0
-  var segmentBufferCount = 0
-  var overflowBufferIndex = 0
-  var overflowBufferCount = 0
+  
+  internal struct NormalizationBuffer {  
+    enum Mode {
+      case stack, heap
+    }
+    var stackStorage: _FixedArray16<CodeUnit>
+    var heapStorage: [CodeUnit] = []
+    var count: Int
+    var readIndex: Int
+    var mode: Mode = .stack
+  
+    init() {
+      stackStorage = _FixedArray16(allZeros: ())
+      count = 0
+      readIndex = 0
+    }
+    
+    mutating func reset() {
+      count = 0
+      readIndex = 0
+    }
+    
+    var isStackBased: Bool {
+      return mode == .stack
+    }
+    
+    var canRead: Bool {
+      return readIndex < count
+    }
+    
+    mutating func read() -> CodeUnit {
+      defer { readIndex += 1 }
+      return self[readIndex]
+    }
+    
+    mutating func promote(_ size: Int, multiplier: Int = 1) {
+      if isStackBased {
+        heapStorage = [CodeUnit](repeating: 0, count: (size + count) * multiplier)
+        for i in 0..<count {
+          heapStorage[i] = stackStorage[i]
+        }
+        mode = .heap
+      } else {
+        fatalError("Promoting a heap array doesn't make sense")
+      }
+    }
+    
+    subscript(index: Int) -> CodeUnit {
+      assert(index < count)
+      assert(index >= 0)
+      if isStackBased {
+        return stackStorage[index]
+      } else {
+        return heapStorage[index]
+      }
+    }
+    
+    mutating func withUnsafeBufferPointer(_ f: (UnsafeBufferPointer<CodeUnit>) throws -> Void) rethrows {
+      if isStackBased {
+        try stackStorage.withUnsafeBufferPointer { arrayBuffer in
+          let slice = arrayBuffer[0..<count]
+          let buffer = UnsafeBufferPointer(rebasing: slice)
+          try f(buffer)
+        }
+      } else {
+        try heapStorage.withUnsafeBufferPointer { arrayBuffer in 
+          let slice = arrayBuffer[0..<count]
+          let buffer = UnsafeBufferPointer(rebasing: slice)
+          try f(buffer)
+        }
+      }
+    }
+    
+    @discardableResult
+    mutating func withUnsafeMutableBufferPointer(_ f: (UnsafeMutableBufferPointer<CodeUnit>) throws -> Int) rethrows -> Int{
+      if isStackBased {
+        count = try stackStorage.withUnsafeMutableBufferPointer { arrayBuffer in
+          return try f(arrayBuffer)
+        }
+      } else {
+        count = try heapStorage.withUnsafeMutableBufferPointer { arrayBuffer in
+          return try f(arrayBuffer)
+        }
+      }
+      return count
+    }
+  }
```

Comment:
As discussed, this should probably be replaced with a more helpful method that will try to fill the stack buffer if possible, or promote and fill the heap buffer. 

```swift
// Comment explaining the contract
mutating func fill(promotionSize: Int, using f: (UMBP<CodeUnit>) throws -> Int?) rethrows -> Int {
  if stackBased { ... try to fill the stack, and promote if we can't }
  ... fill the heap ...
}
```

And then all the callers don't have perform error handling or copy-paste the body of `f` into multiple places (and have their own fatalErrors).